src/aws_encryption_sdk/materials_managers/__init__.py

```diff
@@ -40,38 +48,250 @@ class EncryptionMaterialsRequest(object):
+    :param keyring_trace: Any KeyRing trace entries (optional)
+    :type keyring_trace: list of :class:`KeyringTrace`
     """
 
-    algorithm = attr.ib(validator=attr.validators.instance_of(Algorithm))
-    data_encryption_key = attr.ib(validator=attr.validators.instance_of(DataKey))
-    encrypted_data_keys = attr.ib(validator=attr.validators.instance_of(set))
-    encryption_context = attr.ib(validator=attr.validators.instance_of(dict))
-    signing_key = attr.ib(default=None, validator=attr.validators.optional(attr.validators.instance_of(bytes)))
+    _encrypted_data_keys = attr.ib(
+        default=attr.Factory(list), validator=optional(deep_iterable(member_validator=instance_of(EncryptedDataKey)))
+    )
+    signing_key = attr.ib(default=None, repr=False, validator=optional(instance_of(bytes)))
+
+    def __init__(
+        self,
+        algorithm=None,
+        data_encryption_key=None,
+        encrypted_data_keys=None,
+        encryption_context=None,
+        signing_key=None,
+        **kwargs
+    ):  # noqa we define this in the class docstring
+        if algorithm is None:
+            raise TypeError("algorithm must not be None")
+
+        if encryption_context is None:
+            raise TypeError("encryption_context must not be None")
+
+        super(EncryptionMaterials, self).__init__(
+            algorithm=algorithm,
+            encryption_context=encryption_context,
+            data_encryption_key=data_encryption_key,
+            **kwargs
+        )
+        self._setattr("signing_key", signing_key)
+        self._setattr("_encrypted_data_keys", encrypted_data_keys)
+        attr.validate(self)
+
+    @property
+    def encrypted_data_keys(self):
+        """Return a read-only version of the encrypted data keys.
+
+        :rtype: frozenset
+        """
+        return frozenset(self._encrypted_data_keys)
+
+    def add_data_encryption_key(self, data_encryption_key, keyring_trace):
+        # type: (Union[DataKey, RawDataKey], KeyringTrace) -> None
+        """Add a plaintext data encryption key.
+
+        .. versionadded:: 1.5.0
+
+        :param RawDataKey data_encryption_key: Data encryption key
+        :param KeyringTrace keyring_trace: Trace of actions that a keyring performed
+          while getting this data encryption key
+        :raises AttributeError: if data encryption key is already set
+        :raises InvalidKeyringTraceError: if keyring trace does not match generate action
+        :raises InvalidKeyringTraceError: if keyring trace does not match data key provider
+        :raises InvalidDataKeyError: if data key length does not match algorithm suite
+        """
+        self._add_data_encryption_key(
+            data_encryption_key=data_encryption_key,
+            keyring_trace=keyring_trace,
+            required_flags={KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY},
+        )
+
+    def add_encrypted_data_key(self, encrypted_data_key, keyring_trace):
```

Comment:
I would also add a check to make sure the `data_encryption_key` has been set.
https://github.com/awslabs/aws-encryption-sdk-javascript/blob/master/modules/material-management/src/cryptographic_material.ts#L358

Reply:
Yes, I like that.

Reply:
merged in https://github.com/aws/aws-encryption-sdk-python/pull/163/commits/b311cda30958283860f198f2f61a24d928af6d9e